### ISSUE TITLE
Feature Request: Renderer Dependencies

### ISSUE THREAD

**Issue by harrand** [harrand/Topaz#74]
**Describe the feature request**
Right now, synchronising renderers is extremely barebones. By default, renderers will submit work whenever requested, and that work doesn't wait on anything. There is a RendererOption which causes render() to block until the work has finished, but that's only useful if *everything* needs to stop until some render work is done.

It is common for a renderer to do work that is depended on by another, but we don't want to halt the CPU until the initial work is done. The golden solution to this would be a render graph, but I don't think that's necessary. Instead, renderers (or the device itself) should be able to register dependencies between renderers, like subpass dependencies in vulkan.

Example - Renderer Specifies Dependency:
```cpp
tz::gl::RendererHandle rh = //...
tz::gl::RendererInfo rinfo2;
//...
rinfo2.add_dependency(rh);
```
Example 2 - Device Specifies Dependencies:
```cpp
tz::gl::RendererHandle rh = //...
tz::gl::RendererHandle rh2 = //...
tz::gl::device().add_dependency(rh, rh2); // rh2's render/compute work now waits on a semaphore to be signalled by the completion of rh's work.
```

The former is probably ideal instead of the latter, but implementation details may necessitate the latter. I don't think a render graph should be considered unless neither of these work or it turns out to be super fiddly/difficult in practice.

**Justification**
Render passes depending on each other is ubiquitous, and the lack of this feature forces us to use RenderWait in places where it's not necessary. The biggest example of this is dbgui - right now the dbgui intermediate renderer has RenderWait because it needs to be done by the time the dbgui final renderer invokes. This forces the cpu to wait for each intermediate render pass, which is awful for perf. With this new feature, RenderWait can be removed from that renderer and instead we specify a dependency for the final renderer on the intermediate renderer (note that im not sure how this will work in practice as the intermediate renderer may run several times before the final renderer does, which might cause semaphore shenanigans)

**Additional context**
[zeux.io](https://zeux.io/2020/02/27/writing-an-efficient-vulkan-renderer/) does list some flaws with this approach in the render graph portion, which is worth bearing in mind. Especially as there is a nonzero chance we are forced to use a render graph in the future.

**Comment by harrand** [comment]
API for this now exists at 688066ededfac36462dae15d93fc187efcbc93f4 however there is no implementation.

**Comment by harrand** [comment]
I have been working on an implementation for this for the past few months. It can be found on the [vk_device2](https://github.com/harrand/Topaz/tree/vk_device2) branch.

The following syntax/API was introduced to specify renderer dependencies at [vk_device2 - 289568f](https://github.com/harrand/Topaz/commit/289568f5c63077e68af2445b5d12435cf01fcb71#diff-cd1d2dde85022fb2ce4b91cf0fc9078eec4de6d0acac72ee6aae9e47b3236608R46-R50):
```cpp
tz::gl::renderer2& compute_worker = tz::gl::get_device2().get_renderer(compute_workerh);
tz::gl::renderer2& renderer = tz::gl::get_device2().get_renderer(rendererh);
tz::gl::get_device2().render_graph().timeline = {compute_workerh, rendererh};
tz::gl::get_device2().render_graph().add_dependencies(rendererh, compute_workerh);
```

The timeline represents the expected order of execution of renderers throughout the frame. In this example `rendererh` depends on `compute_workerh`, meaning that synchronisation logic will ensure that the render work for `rendererh` will not begin until the work for `compute_workerh` has completed.

This means that it is now possible to call `renderer.render(...)` wrongly. If you run `render` on renderers not matching the order you specified, you will run into sync errors. To remedy this, a new `tz::gl::device::render()` function has been added. This will invoke `render()` on all the renderers in the timeline in the order you specified in the timeline. Renderer dependencies shall be respected. Obviously if you need to do some cpu work between renderer invocations then you can't use this - in that scenario you will just need to make sure you invoke the renderers in the correct order yourself.

If you ever try to invoke a renderer when the timeline is empty, a warning will be emitted. If it's just a single renderer, the sync logic would've been trivial anyway so unlikely to be any harm. If you expect to run a whole game's render pipeline though without populating the render graph then you are doomed to fail.

Note that this is only implemented for the new vulkan renderer, and this work isn't even merged to master yet.

**Comment by harrand** [comment]
The aforementioned work is now complete, and is available (read: mandatory) in [Topaz 4.0](https://github.com/harrand/Topaz/tree/Topaz4.0). Closing.
